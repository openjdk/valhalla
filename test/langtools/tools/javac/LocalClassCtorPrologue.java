/*
 * @test  /nodynamiccopyright/
 * @bug 8328649
 * @summary Verify local classes in constructor prologues don't have enclosing instances
 * @compile/fail/ref=LocalClassCtorPrologue.out -XDrawDiagnostics LocalClassCtorPrologue.java
 * @enablePreview
 */

class LocalClassCtorPrologue {

    int x;

    LocalClassCtorPrologue() {
        class Local {
            {
                x++;                // this should fail
            }
        }
        super();
    }

    public class Inner {
        public Inner() {
            class Local {
                {
                    x++;            // this should work
                }
            };
            super();
        }
    }

    class Inner2 {
        Inner2() {
            class Sup {
                int x;
            }
            class Local extends Sup {
                Local() {
                    x = 42;   // x is declared in a super type, error
                    super();
                }
            }
            super();
        }
    }

    class Inner3 {
        Inner3() {
            class Local {
                class Foo {
                    int x;

                    Foo() {
                        x = 42;  // this is OK `x` belongs to Foo
                        super();
                    }
                }
            }
            super();
        }
    }
}
