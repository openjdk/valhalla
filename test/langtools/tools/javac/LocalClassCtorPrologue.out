LocalClassCtorPrologue.java:16:17: compiler.err.cant.ref.before.ctor.called: x
LocalClassCtorPrologue.java:40:21: compiler.err.cant.ref.before.ctor.called: x
- compiler.note.preview.filename: LocalClassCtorPrologue.java, DEFAULT
- compiler.note.preview.recompile
2 errors
